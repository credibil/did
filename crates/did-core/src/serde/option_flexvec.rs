use std::fmt;
use std::marker::PhantomData;
use std::str::FromStr;

use serde::de;
use serde::de::{DeserializeOwned, Deserializer, SeqAccess, Visitor};
use serde::ser::{Serialize, SerializeSeq, Serializer};

pub fn serialize<T, S>(value: &Option<Vec<T>>, serializer: S) -> Result<S::Ok, S::Error>
where
    T: Serialize,
    S: Serializer,
{
    // serialize None as empty array
    if value.is_none() {
        return serializer.serialize_none();
    }

    let some_val = value.as_ref().expect("expected value but got none");

    let mut seq = serializer.serialize_seq(Some(some_val.len()))?;
    for e in some_val {
        seq.serialize_element(e)?;
    }
    seq.end()
}

pub fn deserialize<'de, T, D>(deserializer: D) -> Result<Option<Vec<T>>, D::Error>
where
    T: DeserializeOwned + FromStr,
    D: Deserializer<'de>,
{
    struct VisitorImpl<T>(PhantomData<fn() -> Vec<T>>);

    impl<'de, T> Visitor<'de> for VisitorImpl<T>
    where
        T: DeserializeOwned + FromStr,
    {
        type Value = Option<Vec<T>>;

        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
            formatter.write_str("Option<Vec<<T>>")
        }

        // deserialize array to Vec<T>
        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
        where
            A: SeqAccess<'de>,
        {
            // could be mixed array of strings and objects
            let mut deser: Vec<T> = Vec::new();
            while let Some(curr) = seq.next_element::<serde_json::Value>()? {
                match curr {
                    serde_json::Value::String(s) => {
                        let Ok(res) = T::from_str(&s) else {
                            return Err(de::Error::invalid_type(de::Unexpected::Str(&s), &self));
                        };
                        deser.push(res);
                    }
                    serde_json::Value::Object(o) => {
                        let Ok(res) = serde_json::from_value::<T>(serde_json::Value::Object(o))
                        else {
                            return Err(de::Error::invalid_type(de::Unexpected::Map, &self));
                        };
                        deser.push(res);
                    }
                    _ => {
                        return Err(de::Error::custom(
                            "invalid type: cannot deserialize array element",
                        ));
                    }
                }
            }
            Ok(Some(deser))
        }
    }

    deserializer.deserialize_any(VisitorImpl(PhantomData))
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use anyhow::anyhow;
    use serde::{Deserialize, Serialize};
    use serde_json::json;

    use crate::serde::option_flexvec;

    #[derive(Clone, Debug, Default, Deserialize, Serialize)]
    #[serde(default)]
    struct TestData {
        #[serde(with = "option_flexvec")]
        object_array: Option<Vec<Nested>>,

        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(with = "option_flexvec")]
        array: Option<Vec<String>>,

        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(with = "option_flexvec")]
        none: Option<Vec<String>>,
    }

    #[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
    struct Nested {
        n: String,
    }

    impl FromStr for Nested {
        type Err = anyhow::Error;

        fn from_str(_: &str) -> Result<Self, Self::Err> {
            Err(anyhow!("unimplemented"))
        }
    }

    #[test]
    fn option_flexvec() {
        let test_data = TestData {
            object_array: Some(vec![
                Nested {
                    n: "object1".to_string(),
                },
                Nested {
                    n: "object2".to_string(),
                },
            ]),
            array: Some(vec!["item1".to_string(), "item2".to_string()]),
            none: Default::default(),
        };

        // serialize
        let test_json = serde_json::to_value(&test_data).expect("failed to serialize");
        assert_eq!(
            *test_json.get("object_array").expect("expected value but got none"),
            json!([{"n": "object1"}, {"n": "object2"}])
        );
        assert_eq!(
            *test_json.get("array").expect("expected value but got none"),
            json!(["item1", "item2"]),
        );
        assert_eq!(test_json.get("none"), None);

        // deserialize
        let test_de: TestData = serde_json::from_value(test_json).expect("failed to deserialize");
        assert_eq!(test_de.object_array, test_data.object_array);
        assert_eq!(test_de.array, test_data.array);
        assert_eq!(test_de.none, test_data.none);
    }
}
