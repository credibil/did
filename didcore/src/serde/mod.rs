pub mod flexvec_or_single;
pub mod option_flexvec;
pub mod option_flexvec_or_single;
