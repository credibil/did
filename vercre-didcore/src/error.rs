//! # DID Core Errors
//!
//! This module defines the error types used by the DID Core library, including for traits that
//! may be implemented in other crates.

use std::fmt::Display;

use thiserror::Error;

/// Simplify creation of errors with tracing.
///
/// # Example
/// ```
/// use didcore::error::Err;
/// use didcore::{error, Result};
///
/// fn with_msg() -> Result<(), Err> {
///     error!(Err::InvalidRequest, "message: {}", "some message")
/// }
///
/// fn no_msg() -> Result<(), Err> {
///     error!(Err::InvalidRequest)
/// }
/// ```
#[macro_export]
macro_rules! tracerr {
    // with context
    ($code:expr, $($msg:tt)*) => {
        {
        use $crate::error::Context as _;
        tracing::error!($($msg)*);
        return Err($code).context(format!($($msg)*));
        }
    };
    // no context
    ($code:expr) => {
        {
        tracing::error!("{}", $code);
        return Err($code.into());
        }
    }
}

/// Public error type for DID Core.
#[derive(Error, Debug)]
#[error(transparent)]
pub struct Error(#[from] anyhow::Error);

impl Error {
    /// Transfer the error to `OAuth2` compatible format.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        serde_json::json!({
            "error": self.0.root_cause().to_string(),
            "error_description": self.to_string(),
        })
    }

    /// Returns true if `E` is the type held by this error object.
    #[must_use]
    pub fn is(&self, err: Err) -> bool {
        self.0.downcast_ref::<Err>().map_or(false, |e| e == &err)
    }
}

/// Typed errors for DID Core.
#[derive(Clone, Copy, Error, Debug, PartialEq, Eq)]
pub enum Err {
    /// Hash is not a valid SHA-256 hash.
    #[error("invalid_hash")]
    InvalidHash,

    /// Invalid format. (See context for details)
    #[error("invalid_format")]
    InvalidFormat,

    /// Invalid input. Used where a verification fails that is more complex than a simple incorrect
    /// format. (See context for details)
    #[error("invalid_input")]
    InvalidInput,

    /// Invalid key is where the format of the key is incorrect or the cryptographic algorithm
    /// specified by the key is not supported.
    #[error("invalid_key")]
    InvalidKey,

    /// Invalid patch. This is used when a patch for a DID document is verified for consistency.
    #[error("invalid_patch")]
    InvalidPatch,

    /// Key not found. This is in response to parsing a DID Document for a public key and not
    /// finding one for the specified purpose or asking a keyring for a public key and not finding
    /// one for the specified operation.
    #[error("key_not_found")]
    KeyNotFound,

    /// An error was returned from a downstream API
    #[error("api_error")]
    ApiError,

    /// An error occurred trying to deserialize data.
    #[error("deserialization_error")]
    DeserializationError,

    /// An error occurred trying to serialize data.
    #[error("serialization_error")]
    SerializationError,

    /// Environment configuration could not be resolved.
    #[error("invalid_config")]
    InvalidConfig,

    /// Authentication failed.
    #[error("auth_error")]
    AuthError,

    /// Request failed. This is used when a request to a downstream API fails to connect or get a
    /// response.
    #[error("request_error")]
    RequestError,

    /// Failure to sign a message.
    #[error("signing_error")]
    SigningError,

    /// Failure to verify a signature.
    #[error("failed_signature_verification")]
    FailedSignatureVerification,

    /// A requested key signing algorithm is not supported by either the signer or the key store.
    #[error("unsupported_algorithm")]
    UnsupportedAlgorithm,

    /// An expiry date is in the past.
    #[error("expired")]
    Expired,

    /// An unspecified error occurred (see context for information)
    #[error("unknown")]
    Unknown,

    /// Feature is not yet implemented.
    #[error("not_implemented")]
    NotImplemented,

    /// Feature is not supported.
    #[error("not_supported")]
    NotSupported,

    /// No DID document was found for the requested DID.
    #[error("not_found")]
    NotFound,
}

/// Context is used to decorate errors with useful context information.
pub trait Context<T, E>
where
    E: std::error::Error + Send + Sync + 'static,
{
    /// Adds context to the error.
    ///
    /// # Arguments
    ///
    /// * `context` - The context to add to the error.
    ///
    /// # Returns
    ///
    /// Original return object or error with context appended.
    ///
    /// # Errors
    ///
    /// * Original error with context appended.
    fn context<C>(self, context: C) -> Result<T, Error>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> Context<T, E> for core::result::Result<T, E>
where
    E: std::error::Error + Send + Sync + 'static,
{
    fn context<C>(self, context: C) -> Result<T, Error>
    where
        C: Display + Send + Sync + 'static,
    {
        match self {
            Ok(ok) => Ok(ok),
            Err(e) => Err(Error(anyhow::Error::from(e).context(context))),
        }
    }
}

impl From<Err> for Error {
    fn from(error: Err) -> Self {
        Error(error.into())
    }
}

impl From<base64ct::Error> for Error {
    fn from(err: base64ct::Error) -> Error {
        Error(err.into())
    }
}

impl From<ecdsa::Error> for Error {
    fn from(err: ecdsa::Error) -> Error {
        Error(err.into())
    }
}

impl From<multihash::Error> for Error {
    fn from(err: multihash::Error) -> Error {
        Error(err.into())
    }
}

impl From<regex::Error> for Error {
    fn from(err: regex::Error) -> Error {
        Error(err.into())
    }
}

impl From<reqwest::Error> for Error {
    fn from(err: reqwest::Error) -> Error {
        Error(err.into())
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Error {
        Error(err.into())
    }
}

impl From<std::string::FromUtf8Error> for Error {
    fn from(err: std::string::FromUtf8Error) -> Error {
        Error(err.into())
    }
}

impl From<url::ParseError> for Error {
    fn from(err: url::ParseError) -> Error {
        Error(err.into())
    }
}

#[cfg(test)]
mod test {
    use serde_json::json;
    use tracing::Level;
    use tracing_subscriber::FmtSubscriber;

    use super::*;
    use crate::Result;

    #[test]
    fn base_err() {
        let err: Error = Err::InvalidFormat.into();

        assert_eq!(
            err.to_json(),
            json!({"error":"invalid_format","error_description":"invalid_format"})
        );
    }

    #[test]
    fn context_err() {
        let res: Result<()> = Err(Err::InvalidFormat).context("Invalid format description");
        let err = res.expect_err("expected error");

        assert_eq!(
            err.to_json(),
            json!({"error":"invalid_format","error_description":"Invalid format description"})
        );
    }

    #[test]
    fn test_macro() {
        let subscriber = FmtSubscriber::builder().with_max_level(Level::ERROR).finish();
        tracing::subscriber::set_global_default(subscriber).expect("setting subscriber failed");

        let Err(e) = run_macro() else {
            panic!("expected error");
        };

        assert_eq!(e.to_string(), "test me");
    }

    fn run_macro() -> Result<()> {
        tracerr!(Err::InvalidFormat, "test {}", "me")
    }
}
