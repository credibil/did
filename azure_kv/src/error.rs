use serde::Deserialize;

/// Error returned from Azure Key Vault.
#[derive(Debug, Deserialize)]
pub struct ApiErrorResponse {
    /// Error details.
    pub error: ApiErrorDetail,
}

/// Error returned from Azure Key Vault.
#[derive(Debug, Deserialize)]
pub struct ApiErrorDetail {
    /// Error code.
    pub code: String,
    /// Error message.
    pub message: String,
}
