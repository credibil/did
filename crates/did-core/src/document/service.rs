//! Services are used to express ways of communicating with the DID subject or associated entities.
//! Can be any type of service the DID subject wants to advertise, including decentralized identity
//! management services for further discovery, authentication, authorization, or interaction.

use std::collections::HashMap;
use std::convert::Infallible;
use std::str::FromStr;

use serde::ser::SerializeMap;
use serde::{Deserialize, Serialize};

use crate::error::Err;
use crate::serde::flexvec_or_single;
use crate::{tracerr, Result};

/// Service description.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", default)]
pub struct Service {
    /// Identifier for the service. Should be unique for services within the DID document.
    pub id: String,
    /// The type of service.
    #[serde(rename = "type")]
    #[serde(with = "flexvec_or_single")]
    pub type_: Vec<String>,
    /// Location(s) of the service.
    #[serde(with = "endpoint_serialization")]
    pub service_endpoint: Vec<Endpoint>,
}

/// A service endpoint can be a string, a map or a set composed of one or more strings and/or maps.
/// All string values must be valid URIs.
#[derive(Clone, Debug, Default, Deserialize)]
pub struct Endpoint {
    /// Location of the service endpoint, if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    /// Map of endpoint names to locations, if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url_map: Option<HashMap<String, Vec<String>>>,
}

/// Required by serde to deserialise a service endpoint using [`crate::serde::flexvec`].
impl FromStr for Endpoint {
    type Err = Infallible;

    fn from_str(url: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            url: Some(url.to_string()),
            url_map: None,
        })
    }
}

/// Serialize a service endpoint to a string or map. If only the `url` field is set, serialize to a
/// string, otherwise serialize to a map.
impl Serialize for Endpoint {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match (&self.url, &self.url_map) {
            (Some(url), None) => serializer.serialize_str(url),
            (None, Some(map)) => {
                let mut m = serializer.serialize_map(Some(map.len()))?;
                for (k, v) in map {
                    m.serialize_entry(k, v)?;
                }
                m.end()
            }
            _ => Err(serde::ser::Error::custom("Service endpoint must be a string or map")),
        }
    }
}

/// Check the services in a set conform to format constraints.
///
/// # Errors
///
/// - [`Err::InvalidInput`] if a service ID is duplicated
/// - [`Err::InvalidFormat`] if a service type exceeds the limit of 30 characters
/// - URL parsing error if a service endpoint is not a valid URL
pub fn check_services(services: &[Service]) -> Result<()> {
    let mut map = HashMap::new();
    for s in services {
        // Check duplicate
        if map.contains_key(&s.id) {
            tracerr!(Err::InvalidInput, "Duplicate service ID: {}", s.id);
        }
        map.insert(s.id.clone(), true);

        // Check length
        if s.type_.len() > 30 {
            tracerr!(Err::InvalidFormat, "Service type exceeds limit of 30: {}", s.type_.len());
        }

        // Check all endpoint representations are valid URLs
        for ep in &s.service_endpoint {
            if let Some(url) = &ep.url {
                let _ = url::Url::parse(url)?;
            }
            if let Some(map) = &ep.url_map {
                for v in map.values() {
                    for url in v {
                        let _ = url::Url::parse(url)?;
                    }
                }
            }
        }
    }
    Ok(())
}

pub(super) mod endpoint_serialization {
    use std::collections::HashMap;
    use std::fmt;
    use std::marker::PhantomData;
    use std::str::FromStr;

    use serde::de::{self, MapAccess, SeqAccess, Visitor};
    use serde::ser::{SerializeSeq, Serializer};

    use super::{Endpoint, Result};

    pub(crate) fn serialize<S>(value: &[Endpoint], serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        if value.len() == 1 {
            serializer.serialize_some(&value[0])
        } else {
            let mut seq = serializer.serialize_seq(Some(value.len()))?;
            for e in value {
                seq.serialize_element(e)?;
            }
            seq.end()
        }
    }

    #[allow(clippy::too_many_lines)]
    pub(crate) fn deserialize<'de, D>(deserializer: D) -> Result<Vec<Endpoint>, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        struct StringOrMap<T>(PhantomData<fn() -> Vec<T>>);

        impl<'de> Visitor<'de> for StringOrMap<Vec<Endpoint>> {
            type Value = Vec<Endpoint>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str(
                    "ServiceEndpoint as a single string or map or a set of strings and/or maps",
                )
            }

            // If a string is found, deserialize it to a single entry Vec<ServiceEndpoint> using
            // ServiceEndpoint::from_str
            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                match Endpoint::from_str(value) {
                    Ok(res) => Ok(vec![res]),
                    Err(_) => Err(de::Error::invalid_type(de::Unexpected::Str(value), &self)),
                }
            }

            fn visit_map<A>(self, mut access: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'de>,
            {
                let mut deser: HashMap<String, Vec<String>> = HashMap::new();
                while let Some((k, v)) = access.next_entry::<String, serde_json::Value>()? {
                    match v {
                        serde_json::Value::String(s) => {
                            deser.insert(k, vec![s]);
                        }
                        serde_json::Value::Array(a) => {
                            let mut vec: Vec<String> = Vec::new();
                            for e in a {
                                match e {
                                    serde_json::Value::String(s) => {
                                        vec.push(s);
                                    }
                                    _ => {
                                        return Err(de::Error::invalid_type(
                                            de::Unexpected::Map,
                                            &self,
                                        ));
                                    }
                                }
                            }
                            deser.insert(k, vec);
                        }
                        _ => {
                            return Err(de::Error::invalid_type(de::Unexpected::Map, &self));
                        }
                    }
                }

                Ok(vec![Endpoint {
                    url: None,
                    url_map: Some(deser),
                }])
            }

            // Deserialize a sequence to Vec<Context>
            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: SeqAccess<'de>,
            {
                // could be mixed array of strings and objects
                let mut deser: Vec<Endpoint> = Vec::new();
                while let Some(curr) = seq.next_element::<serde_json::Value>()? {
                    match curr {
                        serde_json::Value::String(s) => {
                            let Ok(res) = Endpoint::from_str(&s) else {
                                return Err(de::Error::invalid_type(
                                    de::Unexpected::Str(&s),
                                    &self,
                                ));
                            };
                            deser.push(res);
                        }
                        serde_json::Value::Object(o) => {
                            // Try to deserialize as a map of strings
                            let map_of_vecs = match serde_json::from_value::<HashMap<String, String>>(
                                serde_json::Value::Object(o.clone()),
                            ) {
                                Ok(res) => {
                                    // Convert to map of strings to map of vec of strings
                                    let mut map: HashMap<String, Vec<String>> = HashMap::new();
                                    for (k, v) in res {
                                        map.insert(k, vec![v]);
                                    }
                                    map
                                }
                                Err(_) => {
                                    // Try again with a map of vec of strings directly
                                    match serde_json::from_value::<HashMap<String, Vec<String>>>(
                                        serde_json::Value::Object(o.clone()),
                                    ) {
                                        Ok(res) => res,
                                        Err(_) => {
                                            return Err(de::Error::invalid_type(
                                                de::Unexpected::Map,
                                                &self,
                                            ));
                                        }
                                    }
                                }
                            };
                            let se = Endpoint {
                                url: None,
                                url_map: Some(map_of_vecs),
                            };
                            deser.push(se);
                        }
                        _ => {
                            return Err(de::Error::custom(
                                "invalid type: cannot deserialize array element",
                            ));
                        }
                    }
                }
                Ok(deser)
            }
        }

        deserializer.deserialize_any(StringOrMap(PhantomData))
    }
}

#[cfg(test)]
mod tests {
    use olpc_cjson::CanonicalFormatter;

    use super::*;

    #[derive(Serialize, Deserialize)]
    struct Doc {
        pub service: Vec<Service>,
    }

    #[test]
    fn deserialize_single_string() {
        let input = r#"{
            "service": [
                {
                    "id": "did:example:123456789abcdefghi#openid",
                    "type": ["OpenIdConnectVersion1.0Service"],
                    "serviceEndpoint": "https://openid.example.com/"
                }
            ]
        }"#;
        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_single_map() {
        let input = r#"{
            "service": [
                {
                    "id": "did:example:123456789abcdefghi#openid",
                    "type": [ "OpenIdConnectVersion1.0Service" ],
                    "serviceEndpoint": { "origin": "https://openid.example.com/" }
                }
            ]
        }"#;
        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_strings() {
        let input = r#"{
            "service": [
                {
                    "id": "did:example:123456789abcdefghi#openid",
                    "type": ["OpenIdConnectVersion1.0Service"],
                    "serviceEndpoint": ["https://openid.example.com/", "https://auth.example.com/"]
                }
            ]
        }"#;
        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_maps() {
        let input = r#"{
            "service": [
                {
                    "id": "did:example:123456789abcdefghi#openid",
                    "type": ["OpenIdConnectVersion1.0Service"],
                    "serviceEndpoint": [
                        { "origin": "https://openid.example.com/" },
                        { "alt": ["https://auth.example.com/"] }
                    ]
                }
            ]
        }"#;
        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn deserialize_mix() {
        let input = r#"{
            "service": [
                {
                    "id": "did:example:123456789abcdefghi#openid",
                    "type": ["OpenIdConnectVersion1.0Service"],
                    "serviceEndpoint": [
                        "https://openid.example.com/",
                        { "alt": ["https://auth.example.com/"], "origin": ["https://openid.example.com/"] }
                    ]
                }
            ]
        }"#;
        let doc: Doc = serde_json::from_str(input).expect("failed to deserialize");
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }

    #[test]
    fn serialize_single_string() {
        let doc = Doc {
            service: vec![Service {
                id: "did:example:123456789abcdefghi#openid".to_string(),
                type_: vec!["OpenIdConnectVersion1.0Service".to_string()],
                service_endpoint: vec![Endpoint {
                    url: Some("https://openid.example.com/".to_string()),
                    url_map: None,
                }],
            }],
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let json = String::from_utf8(buf).expect("failed to convert bytes to string");
        assert_eq!(
            json,
            r#"{"service":[{"id":"did:example:123456789abcdefghi#openid","serviceEndpoint":"https://openid.example.com/","type":"OpenIdConnectVersion1.0Service"}]}"#
        );
    }

    #[test]
    fn serialize_single_map() {
        let doc = Doc {
            service: vec![Service {
                id: "did:example:123456789abcdefghi#openid".to_string(),
                type_: vec!["OpenIdConnectVersion1.0Service".to_string()],
                service_endpoint: vec![Endpoint {
                    url: None,
                    url_map: Some(HashMap::from([(
                        "origin".to_string(),
                        vec!["https://openid.example.com/".to_string()],
                    )])),
                }],
            }],
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let json = String::from_utf8(buf).expect("failed to convert bytes to string");
        assert_eq!(
            json,
            r#"{"service":[{"id":"did:example:123456789abcdefghi#openid","serviceEndpoint":{"origin":["https://openid.example.com/"]},"type":"OpenIdConnectVersion1.0Service"}]}"#
        );
    }

    #[test]
    fn serialize_mix() {
        let doc = Doc {
            service: vec![Service {
                id: "did:example:123456789abcdefghi#openid".to_string(),
                type_: vec!["OpenIdConnectVersion1.0Service".to_string()],
                service_endpoint: vec![
                    Endpoint {
                        url: Some("https://openid.example.com/".to_string()),
                        url_map: None,
                    },
                    Endpoint {
                        url: None,
                        url_map: Some(HashMap::from([(
                            "alt".to_string(),
                            vec!["https://auth.example.com/".to_string()],
                        )])),
                    },
                ],
            }],
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let json = String::from_utf8(buf).expect("failed to convert bytes to string");
        assert_eq!(
            json,
            r#"{"service":[{"id":"did:example:123456789abcdefghi#openid","serviceEndpoint":["https://openid.example.com/",{"alt":["https://auth.example.com/"]}],"type":"OpenIdConnectVersion1.0Service"}]}"#
        );
    }
}
