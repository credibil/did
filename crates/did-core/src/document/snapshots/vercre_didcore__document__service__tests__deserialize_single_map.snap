---
source: did-core/src/document/service.rs
assertion_line: 306
expression: doc
---
service:
  - id: "did:example:123456789abcdefghi#openid"
    type: OpenIdConnectVersion1.0Service
    serviceEndpoint:
      origin:
        - "https://openid.example.com/"

